--- a/cube/contents/ui/ScreenView.qml
+++ b/cube/contents/ui/ScreenView.qml
@@ -16,12 +16,25 @@
     focus: true
 
     readonly property QtObject targetScreen: KWinComponents.SceneView.screen
+    property QtObject currentDesktop: {
+        if (typeof KWinComponents.Workspace.currentDesktopForScreen === "function") {
+            return KWinComponents.Workspace.currentDesktopForScreen(root.targetScreen);
+        } else {
+            // Fallback for kwin-x11
+            return KWinComponents.Workspace.currentDesktop;
+        }
+    }
 
     function switchTo(desktop) {
-        KWinComponents.Workspace.currentDesktop = desktop;
+        if (typeof KWinComponents.Workspace.setCurrentDesktopForScreen === "function") {
+            KWinComponents.Workspace.setCurrentDesktopForScreen(desktop, root.targetScreen);
+        } else {
+            // Fallback for kwin-x11
+            KWinComponents.Workspace.currentDesktop = desktop;
+        }
+
         effect.deactivate();
     }
-
     function switchToSelected() {
         const eulerRotation = cameraController.rotation.toEulerAngles();
         switchTo(cube.desktopAt(eulerRotation.y));
@@ -30,7 +43,8 @@
     View3D {
         id: view
         anchors.fill: parent
-        renderMode: View3D.Underlay
+        // TODO: Use the Underlay mode again when the QtQuick3D renderer accounts for QQuickRenderTarget::mirrorVertically().
+        renderMode: View3D.Offscreen
 
         Loader {
             id: transparentSceneEnvironment
@@ -92,7 +106,7 @@
                     PropertyChanges {
                         target: cameraController
                         radius: cube.faceDistance + 0.5 * cube.faceSize.height / Math.tan(0.5 * camera.fieldOfView * Math.PI / 180)
-                        rotation: Quaternion.fromEulerAngles(0, cube.desktopAzimuth(KWinComponents.Workspace.currentDesktop), 0)
+                        rotation: Quaternion.fromEulerAngles(0, cube.desktopAzimuth(currentDesktop), 0)
                     }
                 },
                 State {
@@ -100,7 +114,7 @@
                     PropertyChanges {
                         target: cameraController
                         radius: cube.faceDistance * effect.configuration.DistanceFactor + 0.5 * cube.faceSize.height / Math.tan(0.5 * camera.fieldOfView * Math.PI / 180)
-                        rotation: Quaternion.fromEulerAngles(0, cube.desktopAzimuth(KWinComponents.Workspace.currentDesktop), 0).times(Quaternion.fromEulerAngles(-20, 0, 0))
+                        rotation: Quaternion.fromEulerAngles(0, cube.desktopAzimuth(currentDesktop), 0).times(Quaternion.fromEulerAngles(-20, 0, 0))
                     }
                 }
             ]
@@ -164,4 +178,21 @@
     Keys.onEnterPressed: root.switchToSelected();
     Keys.onReturnPressed: root.switchToSelected();
     Keys.onSpacePressed: root.switchToSelected();
+
+    Connections {
+        target: KWinComponents.Workspace
+        onCurrentDesktopChanged: (previous, current, screen) => {
+            // fallback for kwin-x11
+            if (current === undefined && screen === undefined) {
+                root.currentDesktop = KWinComponents.Workspace.currentDesktop;
+                return;
+            }
+
+            if (screen !== root.targetScreen) {
+                return;
+            }
+
+            root.currentDesktop = current;
+        }
+    }
 }
